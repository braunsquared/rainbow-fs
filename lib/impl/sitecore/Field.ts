import { IField } from '../../model/Sitecore';

export class Field implements IField {
  ID: string = "";
  Hint: string;
  Type?: string;
  Value: any;

  constructor(id: string, hint: string, value: any) {
    this.ID = id;
    this.Hint = hint;
    this.Value = value;
  }

  toObject(): any {
    const obj: any = {
      ID: this.ID,
      Hint: this.Hint,
      Value: this.Value,
    };

    if(this.Type) {
      obj.Type = this.Type;
    }

    return obj;
  }

  static fromObject(obj: any): IField {
    const field = new Field(obj.ID, obj.Hint, obj.Value);
    field.Type = obj.Type;
    return field;
  }
}