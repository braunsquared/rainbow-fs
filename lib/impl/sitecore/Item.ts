import { GUID, IItem, IPath, ILanguage, IField, ItemMetaData } from '../../model/Sitecore';
import { Field } from './Field';
import { Language } from './Language';
import { Path } from './Path';
import { IStore } from '../../Store';

export function defaultItemFactory(store: IStore, obj: any, meta?: ItemMetaData): IItem {
  const itemObj = {...obj};
  if (!itemObj.ID) {
    itemObj.ID = store.generateId(store.config.idSource(itemObj));
  }

  const item = Item.fromObject(itemObj);

  if(meta) {
    item.Meta = {...meta};
  }

  return item;
}

export class Item implements IItem {
  ID: GUID;
  Parent: GUID;
  Template: GUID;
  Path: IPath;
  DB: string = "master";

  SharedFields: Array<IField> = [];
  Languages: Array<ILanguage> = [];

  Meta: ItemMetaData = {
    isNew: true,
    isDirty: false,
    isVirtual: false,
  }

  constructor(id: GUID, path: string) {
    this.ID = id;
    this.Path = new Path(path);
  }

  addSharedField(field: IField): this {
    this.SharedFields.push(field);
    return this;
  }

  removeSharedField(field: IField): this {
    const idx = this.SharedFields.indexOf(field);
    if (idx >= 0) {
      this.SharedFields.splice(idx, 1);
    }
    return this;
  }

  addLanguage(lang: ILanguage): this {
    this.Languages.push(lang);
    return this;
  }

  removeLanguage(lang: ILanguage): this {
    const idx = this.Languages.indexOf(lang);
    if (idx >= 0) {
      this.Languages.splice(idx, 1);
    }
    return this;
  }

  field(idOrName: string, lang: string = 'en'): IField | undefined {
    const field = this.SharedFields.find(f => f.ID === idOrName || f.Hint === idOrName);
    
    if (field) {
      return field;
    }

    const langInst = this.Languages.find(l => l.Language === lang);
    if (langInst) {
      return langInst.field(idOrName);
    }
  }

  toObject(): any {
    const obj: any = {
      ID: this.ID,
      Parent: this.Parent,
      Template: this.Template,
      Path: this.Path.Path,
      DB: this.DB,
    };

    if (this.SharedFields.length) {
      obj.SharedFields = this.SharedFields.map(f => f.toObject());
    }

    if (this.Languages.length) {
      obj.Languages = this.Languages.map(l => l.toObject());
    }

    return obj;
  }

  static fromObject(obj: any): Item {
    const item = new Item(obj.ID, obj.Path);
    item.Parent = obj.Parent;
    item.Template = obj.Template;
    item.DB = obj.DB;

    if (obj.SharedFields) {
      item.SharedFields = obj.SharedFields.map((fieldObj: any) => Field.fromObject(fieldObj));
    }

    if (obj.Languages) {
      item.Languages = obj.Languages.map((langObj: any) => Language.fromObject(langObj));
    }

    return item;
  }
}