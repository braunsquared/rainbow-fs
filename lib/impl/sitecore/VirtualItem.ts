import { Item } from './Item';
import { IField, ILanguage } from '../../model/Sitecore';

export class VirtualItem extends Item {
  Meta = {
    isNew: false,
    isDirty: false,
    isVirtual: true,
  }

  addSharedField(field: IField): this {
    throw new Error('Unable to perform action on virtual item.');
  }

  removeSharedField(field: IField): this {
    throw new Error('Unable to perform action on virtual item.');
  }

  addLanguage(lang: ILanguage): this {
    throw new Error('Unable to perform action on virtual item.');
  }

  removeLanguage(lang: ILanguage): this {
    throw new Error('Unable to perform action on virtual item.');
  }
}