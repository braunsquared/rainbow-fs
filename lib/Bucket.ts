import fs from 'fs-extra';
import yaml from 'js-yaml';
import fsPath from 'path';
import glob from 'glob';
import util from 'util';

import { IStore } from './Store';
import { Path } from './impl/sitecore/Path';
import { VirtualItem } from './impl/sitecore/VirtualItem';
import { HintFilename, IItem, IPath } from './model/Sitecore';

export type ErrorCallbackValue = Error | null | undefined;

const readFile = util.promisify<string, Buffer>(fs.readFile);
const writeFile = util.promisify<string, string>(fs.writeFile);
const mkdirp = util.promisify<string>(fs.mkdirp);
const asyncGlob = util.promisify<string, string[]>(glob);

export class Bucket{
  Filepath: string;
  Path: Path;
  DB: string;

  constructor(filepath: string, sitecorePath: string, db: string) {
    this.Filepath = filepath;
    this.Path = new Path(sitecorePath);
    this.DB = db;
  }

  convertItemPath(scPath: IPath): string {
    const deltaPath = new Path(scPath.Folder.substring(this.Path.Path.length));
    return fsPath.join(this.Filepath, deltaPath.Path, `${scPath.Name}.yml`);
  }

  async read(store: IStore) {
    if (!this.Filepath) {
      throw new Error("Filepath is not set. Unable to read from file system.");
    }

    const files = await asyncGlob(fsPath.join(this.Filepath, '**/*.yml'));

    for(let i = 0; i < files.length; ++i) {
      const file = files[i];
      const buffer: Buffer = await readFile(file);
      const content = yaml.safeLoad(buffer.toString('utf8'), {filename: file});
      const item = store.createItemFromObject(content, {
        isNew: false,
        isDirty: false,
        hints: new Map([[HintFilename, file]]),
      });

      if(item.DB !== this.DB) {
        console.warn('Found item in bucket from different DB');
        return;
      }

      if(item.Path.Folder === this.Path.Path && item.Parent) {
        // We may be able to infer some information about our item
        const db = store.getDatabase(item.DB);
        if(!db) {
          throw new Error('Database should not be null');
        }

        let bucketItem = db.getItem(item.Parent);
        if (!bucketItem) {
          bucketItem = new VirtualItem(item.Parent, this.Path.Path);
          db.addItem(bucketItem);
        }
      }
    }
  }

  async write(item: IItem) {
    if(item.Meta.isVirtual) {
      throw new Error('Unable to serialize virtual items');
    } else if(!item.Path.Path.startsWith(this.Path.Path)) {
      throw new Error('Item is not a descendant of this bucket');
    } else if(item.DB !== this.DB) {
      throw new Error('Item DB does not match the Bucket DB');
    }

    // Figure out what the path should be
    const itemPath = this.convertItemPath(item.Path);
    const basePath = fsPath.dirname(itemPath);

    await mkdirp(basePath);
    await writeFile(itemPath, yaml.safeDump(item.toObject()));
  }
}
