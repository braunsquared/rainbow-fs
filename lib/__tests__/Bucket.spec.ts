import mockfs from 'mock-fs';
import fs from 'fs';
import { Store } from '../Store';
import { Bucket } from "../Bucket";
import { Item } from '../impl/sitecore/Item';
import { loadMockFs } from '../../test/util';

function itemToString(item: Item, indent: number = 0, output: string = ''): string {
  // let next = output;
  // next += Array(indent + 1).join(' ') + item.Path.Path + os.EOL;
  // item.children.forEach(child => {
  //   next = itemToString(child, indent + 2, next);
  // });
  // return next;
  return '';
}

describe('Bucket', () => {
  let bucket: Bucket;
  let store: Store;
  let sampleData: any;

  beforeAll(() => {
    sampleData = loadMockFs();
  });

  beforeEach(() => {
    mockfs(sampleData);
    store = new Store();
    bucket = new Bucket('/data/Templates', '/sitecore/templates/Helix/Project', 'master');
  });

  afterEach(() => {
    mockfs.restore();
  })

  it('should read a directory', () => {
    return bucket.read(store).then(() => {
      mockfs.restore();
      expect(store).toMatchSnapshot();
    });
  });

  it('should fail to read when no file path is set', () => {
    bucket.Filepath = '';
    expect(bucket.read(store)).rejects.toThrow();
  });

  it('should write an updated item to disk', () => {
    return bucket.read(store).then(() => {
      const db = store.getDatabase('master');
      expect(db).not.toBeNull();

      const item = db!.getItem('261918c5-521a-422d-ae59-23f1d39ee0a3');
      expect(item).not.toBeNull();

      return bucket.write(item!).then(() => {
        const contents = fs.readFileSync(bucket.convertItemPath(item!.Path), {encoding: 'utf8'});
        mockfs.restore();
        expect(contents).toMatchSnapshot();
      });
    });
  });

  it('should fail to write a virtual item to disk', () => {
    return bucket.read(store).then(() => {
      const db = store.getDatabase('master');
      expect(db).not.toBeNull();

      const item = db!.getItem('0725a327-cd61-14f8-8aed-8d74b9917c6c');
      expect(item).not.toBeNull();

      expect(item!.Meta.isVirtual).toEqual(true);
      return expect(bucket.write(item!)).rejects.toThrow();
    });
  });

  it('should fail to write a non child item to disk', () => {
    return bucket.read(store).then(() => {
      const other = new Bucket('/data/Renderings', '/sitecore/layout/Renderings/Project/TestProject', 'master');
      return other.read(store).then(() => {
        const db = store.getDatabase('master');
        expect(db).not.toBeNull();

        const item = db!.getItem('87120214-1be8-7c76-05c1-eaebfbffec1d');
        expect(item).not.toBeNull();
        return expect(bucket.write(item!)).rejects.toThrow();
      });
    });
  });

  it('should fail to write an item from a different DB', () => {
    return bucket.read(store).then(() => {
      const db = store.getDatabase('master');
      expect(db).not.toBeNull();

      const item = db!.getItem('261918c5-521a-422d-ae59-23f1d39ee0a3');
      expect(item).not.toBeNull();

      item!.DB = 'core';

      return expect(bucket.write(item!)).rejects.toThrow();
    });
  });
});