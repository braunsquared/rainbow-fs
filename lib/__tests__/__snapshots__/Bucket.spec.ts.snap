// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bucket should read a directory 1`] = `
Store {
  "_config": Object {
    "dbFactory": [Function],
    "idSource": [Function],
    "itemFactory": [Function],
  },
  "_dbs": Map {
    "master" => MemoryDb {
      "_items": Map {
        "228d3c60-152b-c3e3-f136-684c4d06d061" => Item {
          "DB": "master",
          "ID": "228d3c60-152b-c3e3-f136-684c4d06d061",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20170511T175239Z",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "0725a327-cd61-14f8-8aed-8d74b9917c6c",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project",
            "Name": "TestProject",
            "Path": "/sitecore/templates/Helix/Project/TestProject",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
            ],
          },
          "SharedFields": Array [],
          "Template": "0437fee2-44c9-46a6-abe9-28858d9fee8c",
        },
        "0725a327-cd61-14f8-8aed-8d74b9917c6c" => VirtualItem {
          "DB": "master",
          "ID": "0725a327-cd61-14f8-8aed-8d74b9917c6c",
          "Languages": Array [],
          "Meta": Object {
            "isDirty": false,
            "isNew": false,
            "isVirtual": true,
          },
          "Path": Path {
            "Folder": "/sitecore/templates/Helix",
            "Name": "Project",
            "Path": "/sitecore/templates/Helix/Project",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
            ],
          },
          "SharedFields": Array [],
        },
        "2b43feed-1143-44d1-96f8-c2ee715f2102" => Item {
          "DB": "master",
          "ID": "2b43feed-1143-44d1-96f8-c2ee715f2102",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180919T211302Z",
                    },
                    Field {
                      "Hint": "__Created by",
                      "ID": "5dd74568-4d4b-44c1-b513-0af5f4cda34f",
                      "Type": undefined,
                      "Value": "sitecore\\\\admin
",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Components.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "228d3c60-152b-c3e3-f136-684c4d06d061",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject",
            "Name": "Components",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Components",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Components",
            ],
          },
          "SharedFields": Array [],
          "Template": "0437fee2-44c9-46a6-abe9-28858d9fee8c",
        },
        "261918c5-521a-422d-ae59-23f1d39ee0a3" => Item {
          "DB": "master",
          "ID": "261918c5-521a-422d-ae59-23f1d39ee0a3",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180919T211849Z",
                    },
                    Field {
                      "Hint": "__Created by",
                      "ID": "5dd74568-4d4b-44c1-b513-0af5f4cda34f",
                      "Type": undefined,
                      "Value": "sitecore\\\\admin
",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Components\\\\ContentBlock.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "2b43feed-1143-44d1-96f8-c2ee715f2102",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Components",
            "Name": "ContentBlock",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Components",
              "ContentBlock",
            ],
          },
          "SharedFields": Array [
            Field {
              "Hint": "__Icon",
              "ID": "06d5295c-ed2f-4a54-9bf2-26228d113318",
              "Type": undefined,
              "Value": "office/32x32/document_tag.png",
            },
            Field {
              "Hint": "__Base template",
              "ID": "12c33f3f-86c5-43a5-aeb4-5598cec45116",
              "Type": "tree list",
              "Value": "{1930BBEB-7805-471A-A3BE-4858AC7CF696}",
            },
            Field {
              "Hint": "__Standard values",
              "ID": "f7d48a55-2158-4f02-9356-756654404f73",
              "Type": undefined,
              "Value": "{E4064C34-B165-4916-9081-ECD6970661D1}",
            },
          ],
          "Template": "ab86861a-6030-46c5-b394-e8f99e8b87db",
        },
        "e4064c34-b165-4916-9081-ecd6970661d1" => Item {
          "DB": "master",
          "ID": "e4064c34-b165-4916-9081-ecd6970661d1",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180919T212053Z",
                    },
                    Field {
                      "Hint": "__Created by",
                      "ID": "5dd74568-4d4b-44c1-b513-0af5f4cda34f",
                      "Type": undefined,
                      "Value": "sitecore\\\\admin
",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Components\\\\ContentBlock\\\\__Standard Values.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "261918c5-521a-422d-ae59-23f1d39ee0a3",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock",
            "Name": "__Standard Values",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock/__Standard Values",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Components",
              "ContentBlock",
              "__Standard Values",
            ],
          },
          "SharedFields": Array [],
          "Template": "261918c5-521a-422d-ae59-23f1d39ee0a3",
        },
        "1bc04f33-264f-4c03-9d23-87cf9b14044b" => Item {
          "DB": "master",
          "ID": "1bc04f33-264f-4c03-9d23-87cf9b14044b",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180919T212017Z",
                    },
                    Field {
                      "Hint": "__Created by",
                      "ID": "5dd74568-4d4b-44c1-b513-0af5f4cda34f",
                      "Type": undefined,
                      "Value": "sitecore\\\\admin
",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Components\\\\ContentBlock\\\\Data.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "261918c5-521a-422d-ae59-23f1d39ee0a3",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock",
            "Name": "Data",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock/Data",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Components",
              "ContentBlock",
              "Data",
            ],
          },
          "SharedFields": Array [],
          "Template": "e269fbb5-3750-427a-9149-7aa950b49301",
        },
        "aad95a88-cecc-4f18-ad44-2fdd0f2777cc" => Item {
          "DB": "master",
          "ID": "aad95a88-cecc-4f18-ad44-2fdd0f2777cc",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180919T212018Z",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Components\\\\ContentBlock\\\\Data\\\\content.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "1bc04f33-264f-4c03-9d23-87cf9b14044b",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock/Data",
            "Name": "content",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock/Data/content",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Components",
              "ContentBlock",
              "Data",
              "content",
            ],
          },
          "SharedFields": Array [
            Field {
              "Hint": "Type",
              "ID": "ab162cc0-dc80-4abf-8871-998ee5d7ba32",
              "Type": undefined,
              "Value": "Rich Text",
            },
            Field {
              "Hint": "__Sortorder",
              "ID": "ba3f86a2-4a1c-4d78-b63d-91c2779c1b5e",
              "Type": undefined,
              "Value": 200,
            },
          ],
          "Template": "455a3e98-a627-4b40-8035-e683a0331ac7",
        },
        "5813fa39-94c0-45b4-95fc-f22236aeb9a6" => Item {
          "DB": "master",
          "ID": "5813fa39-94c0-45b4-95fc-f22236aeb9a6",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180919T212018Z",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Components\\\\ContentBlock\\\\Data\\\\heading.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "1bc04f33-264f-4c03-9d23-87cf9b14044b",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock/Data",
            "Name": "heading",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Components/ContentBlock/Data/heading",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Components",
              "ContentBlock",
              "Data",
              "heading",
            ],
          },
          "SharedFields": Array [
            Field {
              "Hint": "Type",
              "ID": "ab162cc0-dc80-4abf-8871-998ee5d7ba32",
              "Type": undefined,
              "Value": "Single-Line Text",
            },
          ],
          "Template": "455a3e98-a627-4b40-8035-e683a0331ac7",
        },
        "f7b28237-9568-41a4-bcc6-3375e24a1a46" => Item {
          "DB": "master",
          "ID": "f7b28237-9568-41a4-bcc6-3375e24a1a46",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180919T211314Z",
                    },
                    Field {
                      "Hint": "__Created by",
                      "ID": "5dd74568-4d4b-44c1-b513-0af5f4cda34f",
                      "Type": undefined,
                      "Value": "sitecore\\\\admin
",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Routes.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "228d3c60-152b-c3e3-f136-684c4d06d061",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject",
            "Name": "Routes",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Routes",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Routes",
            ],
          },
          "SharedFields": Array [],
          "Template": "a87a00b1-e6db-45ab-8b54-636fec3b5523",
        },
        "f529a584-cec1-41a7-9eeb-334938a98285" => Item {
          "DB": "master",
          "ID": "f529a584-cec1-41a7-9eeb-334938a98285",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180917T192424Z",
                    },
                    Field {
                      "Hint": "__Created by",
                      "ID": "5dd74568-4d4b-44c1-b513-0af5f4cda34f",
                      "Type": undefined,
                      "Value": "sitecore\\\\admin
",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Routes\\\\App Route.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "f7b28237-9568-41a4-bcc6-3375e24a1a46",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Routes",
            "Name": "App Route",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Routes/App Route",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Routes",
              "App Route",
            ],
          },
          "SharedFields": Array [
            Field {
              "Hint": "__Icon",
              "ID": "06d5295c-ed2f-4a54-9bf2-26228d113318",
              "Type": undefined,
              "Value": "Apps/16x16/routes.png",
            },
            Field {
              "Hint": "__Base template",
              "ID": "12c33f3f-86c5-43a5-aeb4-5598cec45116",
              "Type": "tree list",
              "Value": "{B36BA9FD-0DC0-49C8-BEA2-E55D70E6AF28}",
            },
            Field {
              "Hint": "__Standard values",
              "ID": "f7d48a55-2158-4f02-9356-756654404f73",
              "Type": undefined,
              "Value": "{8A0619DE-2EE0-4CF3-801C-8CEBEC5D2104}",
            },
          ],
          "Template": "ab86861a-6030-46c5-b394-e8f99e8b87db",
        },
        "8a0619de-2ee0-4cf3-801c-8cebec5d2104" => Item {
          "DB": "master",
          "ID": "8a0619de-2ee0-4cf3-801c-8cebec5d2104",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180917T194639Z",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Routes\\\\App Route\\\\__Standard Values.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "f529a584-cec1-41a7-9eeb-334938a98285",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Routes/App Route",
            "Name": "__Standard Values",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Routes/App Route/__Standard Values",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Routes",
              "App Route",
              "__Standard Values",
            ],
          },
          "SharedFields": Array [
            Field {
              "Hint": "__Renderings",
              "ID": "f1a1fe9e-a60c-4ddb-a3a0-bb5b29fe732e",
              "Type": "layout",
              "Value": "<r xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\">
  <d
    id=\\"{FE5D7FDF-89C0-4D99-9AA3-B5FBD009C9F3}\\"
    l=\\"{C726B750-60D7-49CF-A39C-5C729ACF8531}\\">
    <p
      uid=\\"{42FCCB74-B2B0-4372-85E5-20E0F0A0216E}\\"
      key=\\"jss-main\\"
      md=\\"{174A907F-0627-4691-BF12-3083B65743E3}\\" />
  </d>
</r>
",
            },
          ],
          "Template": "f529a584-cec1-41a7-9eeb-334938a98285",
        },
        "66d58577-b9cb-4b00-a08f-9024fe780b05" => Item {
          "DB": "master",
          "ID": "66d58577-b9cb-4b00-a08f-9024fe780b05",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180917T192504Z",
                    },
                    Field {
                      "Hint": "__Created by",
                      "ID": "5dd74568-4d4b-44c1-b513-0af5f4cda34f",
                      "Type": undefined,
                      "Value": "sitecore\\\\admin
",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Routes\\\\App Route\\\\Page Metadata.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "f529a584-cec1-41a7-9eeb-334938a98285",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Routes/App Route",
            "Name": "Page Metadata",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Routes/App Route/Page Metadata",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Routes",
              "App Route",
              "Page Metadata",
            ],
          },
          "SharedFields": Array [],
          "Template": "e269fbb5-3750-427a-9149-7aa950b49301",
        },
        "5330c186-146f-4145-97e1-10a974493ea4" => Item {
          "DB": "master",
          "ID": "5330c186-146f-4145-97e1-10a974493ea4",
          "Languages": Array [
            Language {
              "Fields": Array [],
              "Language": "en",
              "Versions": Array [
                Version {
                  "Fields": Array [
                    Field {
                      "Hint": "__Created",
                      "ID": "25bed78c-4957-4165-998a-ca1b52f67497",
                      "Type": undefined,
                      "Value": "20180917T192504Z",
                    },
                  ],
                  "Version": 1,
                },
              ],
            },
          ],
          "Meta": Object {
            "hints": Map {
              Symbol(filename) => "C:\\\\data\\\\Templates\\\\TestProject\\\\Routes\\\\App Route\\\\Page Metadata\\\\pageTitle.yml",
            },
            "isDirty": false,
            "isNew": false,
          },
          "Parent": "66d58577-b9cb-4b00-a08f-9024fe780b05",
          "Path": Path {
            "Folder": "/sitecore/templates/Helix/Project/TestProject/Routes/App Route/Page Metadata",
            "Name": "pageTitle",
            "Path": "/sitecore/templates/Helix/Project/TestProject/Routes/App Route/Page Metadata/pageTitle",
            "Tokens": Array [
              "sitecore",
              "templates",
              "Helix",
              "Project",
              "TestProject",
              "Routes",
              "App Route",
              "Page Metadata",
              "pageTitle",
            ],
          },
          "SharedFields": Array [
            Field {
              "Hint": "Type",
              "ID": "ab162cc0-dc80-4abf-8871-998ee5d7ba32",
              "Type": undefined,
              "Value": "Single-Line Text",
            },
          ],
          "Template": "455a3e98-a627-4b40-8035-e683a0331ac7",
        },
      },
      "_ops": Array [],
      "_proxyHandler": ItemProxyHandler {
        "db": [Circular],
      },
      "name": "master",
    },
  },
  "_trees": Map {},
}
`;

exports[`Bucket should write an updated item to disk 1`] = `
"ID: 261918c5-521a-422d-ae59-23f1d39ee0a3
Parent: 2b43feed-1143-44d1-96f8-c2ee715f2102
Template: ab86861a-6030-46c5-b394-e8f99e8b87db
Path: /sitecore/templates/Helix/Project/TestProject/Components/ContentBlock
DB: master
SharedFields:
  - ID: 06d5295c-ed2f-4a54-9bf2-26228d113318
    Hint: __Icon
    Value: office/32x32/document_tag.png
  - ID: 12c33f3f-86c5-43a5-aeb4-5598cec45116
    Hint: __Base template
    Value: '{1930BBEB-7805-471A-A3BE-4858AC7CF696}'
    Type: tree list
  - ID: f7d48a55-2158-4f02-9356-756654404f73
    Hint: __Standard values
    Value: '{E4064C34-B165-4916-9081-ECD6970661D1}'
Languages:
  - Language: en
    Versions:
      - Version: 1
        Fields:
          - ID: 25bed78c-4957-4165-998a-ca1b52f67497
            Hint: __Created
            Value: 20180919T211849Z
          - ID: 5dd74568-4d4b-44c1-b513-0af5f4cda34f
            Hint: __Created by
            Value: |
              sitecore\\\\admin
"
`;
